### PR TITLE
Create executable buffer at the appropriate memory space on GPUs

### PR DESCRIPTION
**Pull request by jaro-sevcik**

### CODE REVIEW

**Comment by jyingl3** [comment]
cc @yashk2810 and @ezhulenev

**Review by jyingl3 — APPROVED** [review]

File: xla/pjrt/pjrt_stream_executor_client.cc [hunk 1]
```diff
@@ -2240,9 +2241,20 @@ std::unique_ptr<PjRtBuffer> OutputBufferHelper(
   std::shared_ptr<TrackedDeviceBuffer> out_buffer =
       TrackedDeviceBuffer::FromScopedShapedBuffer(result_buffer,
                                                   {definition_event});
+  Shape shape = result_buffer->on_device_shape();
+  PjRtMemorySpace* memory_space =
+      device->default_memory_space().value_or(nullptr);
+  if (shape.has_layout() &&
+      shape.layout().memory_space() == Layout::kHostMemorySpace) {
+    absl::StatusOr<PjRtMemorySpace*> memory_space_or =
+        device->memory_space_by_kind(PinnedHostMemorySpace::kKind);
```
Comment (yliu120) [inline xla/pjrt/pjrt_stream_executor_client.cc:2250]: Please use memory_space_by_kind_id whenever possible to accelerate lookup.

File: xla/pjrt/pjrt_stream_executor_client.cc [hunk 2]
```diff
@@ -2240,9 +2241,20 @@ std::unique_ptr<PjRtBuffer> OutputBufferHelper(
...
+    if (memory_space_or.ok()) {
```
Comment (yliu120) [inline xla/pjrt/pjrt_stream_executor_client.cc:2251]: This code looks a bit weird.

I feel like if the lookup failed, you will silently use the default_memory_space without any error or warning.

File: xla/pjrt/pjrt_stream_executor_client.cc [hunk 3]
```diff
@@ -2240,9 +2241,20 @@ std::unique_ptr<PjRtBuffer> OutputBufferHelper(
   std::shared_ptr<TrackedDeviceBuffer> out_buffer =
       TrackedDeviceBuffer::FromScopedShapedBuffer(result_buffer,
                                                   {definition_event});
+  Shape shape = result_buffer->on_device_shape();
```
Comment (yliu120) [inline xla/pjrt/pjrt_stream_executor_client.cc:2244]: const Shape&

File: xla/pjrt/pjrt_stream_executor_client.cc [hunk 4]
```diff
@@ -2240,9 +2241,20 @@ std::unique_ptr<PjRtBuffer> OutputBufferHelper(
...
+  PjRtMemorySpace* memory_space =
+      device->default_memory_space().value_or(nullptr);
```
Comment (yliu120) [inline xla/pjrt/pjrt_stream_executor_client.cc:2246]: https://cs.opensource.google/tensorflow/tensorflow/+/master:third_party/xla/xla/pjrt/gpu/se_gpu_pjrt_client.cc;drc=9e0957281ee29229a2f907cad196a041c1b126d8;l=1144

So junwhan@ made it always returning a value. No need to .value_or(nullptr)?

Comment (jaro-sevcik) [reply xla/pjrt/pjrt_stream_executor_client.cc:2246]: As far as I can see, the interpreter device can still return error, see https://github.com/openxla/xla/blob/60c9ecf92554fe5676b0358fecf93d13863f0476/xla/pjrt/interpreter_device.h#L26, https://github.com/openxla/xla/blob/60c9ecf92554fe5676b0358fecf93d13863f0476/xla/pjrt/pjrt_stream_executor_client.cc#L1364

No?

**Pull request closed without merging**